drivers/interrupt_controller/intc_arcv2_irq_unit.c

```diff
@@ -20,44 +20,112 @@
 #include <arch/cpu.h>
 #include <device.h>
 #include <init.h>
+#include <sw_isr_table.h>
+
+#if defined(CONFIG_ARC_SECURE_FIRMWARE)
+#define _ARC_IRQ_ADJUST_PRIO_S(prio)								\
+	((prio) < ARC_N_IRQ_START_LEVEL ? (prio) : (ARC_N_IRQ_START_LEVEL - 1))
+#define _ARC_IRQ_SET_SECURE(prio)	((prio) | _ARC_V2_IRQ_PRIORITY_SECURE)
+#define ARC_IRQ_PRIO_TO_HW(prio)	_ARC_IRQ_SET_SECURE(_ARC_IRQ_ADJUST_PRIO_S(prio))
+#elif defined(CONFIG_ARC_NORMAL_FIRMWARE)
+#define _ARC_IRQ_ADJUST_PRIO_N(prio)								\
+	((prio) < ARC_N_IRQ_START_LEVEL ? ARC_N_IRQ_START_LEVEL : (prio))
+#define ARC_IRQ_PRIO_TO_HW(prio)	_ARC_IRQ_ADJUST_PRIO_N(prio)
+#else
+#define ARC_IRQ_PRIO_TO_HW(prio)	(prio)
+#endif
+
+#ifdef CONFIG_ARC_CONNECT
+static void arc_shared_intc_init(void)
+{
+	/*
+	 * Initialize all IDU interrupts:
+	 * - select round-robbin
+	 * - disable all lines
+	 */
+	BUILD_ASSERT(CONFIG_NUM_IRQS > ARC_CONNECT_IDU_IRQ_START);
+	__ASSERT(z_arc_v2_core_id() == 0, "idu interrupts must be inited from master core");
+
+	z_arc_connect_idu_disable();
+
+	for (uint32_t i = 0; i < (CONFIG_NUM_IRQS - ARC_CONNECT_IDU_IRQ_START); i++) {
+		/*
+		 * TODO: don't use z_arc_connect_idu* finctions to avoid
+		 * locking/unlocking every time.
+		 */
+		z_arc_connect_idu_set_mode(i, ARC_CONNECT_INTRPT_TRIGGER_LEVEL,
+			ARC_CONNECT_DISTRI_MODE_ROUND_ROBIN);
+		z_arc_connect_idu_set_dest(i, GENMASK(CONFIG_MP_NUM_CPUS, 0));
+
+		/* Disable (mask) line */
+		z_arc_connect_idu_set_mask(i, 0x1);
+	}
+
+	z_arc_connect_idu_enable();
+
+}
+#else
+#define arc_shared_intc_init()
+#endif /* CONFIG_ARC_CONNECT */
+
+static inline void arc_core_intc_init_nolock(uint32_t irq, uint32_t state)
+{
+	z_arc_v2_aux_reg_write(_ARC_V2_IRQ_SELECT, irq);
+	z_arc_v2_aux_reg_write(_ARC_V2_IRQ_PRIORITY,
+		ARC_IRQ_PRIO_TO_HW(_irq_priority_table[irq - CONFIG_GEN_IRQ_START_VECTOR]));
+	z_arc_v2_aux_reg_write(_ARC_V2_IRQ_TRIGGER, _ARC_V2_INT_LEVEL);
+	z_arc_v2_aux_reg_write(_ARC_V2_IRQ_ENABLE, state);
+}
 
 /*
- * @brief Initialize the interrupt unit device driver
+ * Initialize the core private interrupt controller.
  *
- * Initializes the interrupt unit device driver and the device
- * itself.
+ * This function must be called on each CPU in case of SMP system.
  *
- * Interrupts are still locked at this point, so there is no need to protect
- * the window between a write to IRQ_SELECT and subsequent writes to the
- * selected IRQ's registers.
- *
- * @return 0 for success
+ * NOTE: core interrupts are still locked at this point, so there is no need to protect the window
```

Comment:
"interrupts are still locked"? Maybe "masked" or even disabled? And do you mean they are disabled individually or global `STATUS32.IE` flag is not set yet?
For the record, IIRC all IRQs are enabled by default in the private ARC INTC.

Reply:
The interrupts are globally disabled (STATUS32.IE = 0).
I've updated the comment.